File: 2255-minimum-swaps-to-group-all-1s-together-ii/2255-minimum-swaps-to-group-all-1s-together-ii.kt
class Solution {
    fun minSwaps(nums: IntArray): Int {
        var swap_cnt : Int = Int.MAX_VALUE;
        val tot_one = nums.sum();

        var cnt_one : Int = nums[0];
        var e : Int = 0;
        for (st in 0..nums.size) {
            if (st != 0) cnt_one -= nums[st - 1];
            while (e - st + 1 < tot_one) cnt_one += nums[++e % nums.size]
            swap_cnt = min(swap_cnt, tot_one - cnt_one);
        }
        return swap_cnt;
    }
}

File: 2255-minimum-swaps-to-group-all-1s-together-ii/2255-minimum-swaps-to-group-all-1s-together-ii.c
int min(int a, int b) {
    return a < b ? a : b;
}

int minSwaps(int* nums, int numsSize) {
    int swap_cnt = INT_MAX;
    int tot_one = 0;
    for (int i = 0; i < numsSize; i++) tot_one += nums[i];

    int cnt_one = nums[0], e = 0;
    for (int st = 0; st < numsSize; st++) {
        if (st) cnt_one -= nums[st - 1];
        while (e - st + 1 < tot_one) {
            e++;
            cnt_one += nums[e % numsSize];
        }
        swap_cnt = min(swap_cnt, tot_one - cnt_one);
    }
    return swap_cnt;
}